### PR TITLE
multiifo coinc rate module

### PR DESCRIPTION
**Pull request by GarethDaviesGW**
introduction of a coinc_rate module to include the estimation of noise coincidence rates from single trigger rates

### CODE REVIEW

**Comment by tdent** [comment]
@GarethDaviesGW - can you remove the work in progress tag when this is ready to review and assign to Alex N.

File: pycbc/events/coinc_rate.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
+#
+# =============================================================================
+#
+#                                   Preamble
+#
+# =============================================================================
+#
+""" This modules contains functions for calculating and manipulating
```
Comment (tdent) [inline pycbc/events/coinc_rate.py]: modules -> module 
i think 'calculating expected rates of noise and signal coincidences' would be more accurate

File: pycbc/events/coinc_rate.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+def multiifo_noise_coinc_rate(rates, ifos, slop):
+    """
+    Calculate the expected rate of noise coincidences for multiple detectors
+
+    Parameters
+    ----------
+    rates: list of lists/arrays
+        List of vector of single-detector trigger rates
+        in order of template id
```
Comment (tdent) [inline pycbc/events/coinc_rate.py]: do they have to be in order of anything specific?

File: pycbc/events/coinc_rate.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    ifos: list of strings
+        list of ifos in e.g. 'H1' form
```
Comment (tdent) [inline pycbc/events/coinc_rate.py]: I'm guessing that it needs to get the list of ifos in the same order as the list of trigger rates.  if so, would it work better to use a dictionary as input (ie keyed on the ifo name) ?

File: pycbc/events/coinc_rate.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        value is the expected coincidence rate in the combination given the
+        individual detector trigger rates
+    """
+    ifos = numpy.array(ifos)
+    rates = numpy.array(rates)
+    expected_coinc_rates = {}
+    n_ifos = len(ifos)
+    assert len(rates) == n_ifos
+    # Calculate coincidence for all-ifo combination
+    # multiply the two rates and by the overlap time
```
Comment (tdent) [inline pycbc/events/coinc_rate.py]: two or n?

File: pycbc/events/coinc_rate.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    allowed_area = multiifo_noise_coincident_area(ifos, slop)
+    rateprod = [numpy.prod(rs) for rs in zip(*rates)]
+    ifostring = ''.join(ifos)
```
Comment (tdent) [inline pycbc/events/coinc_rate.py]: in other places i think we are using a different layout for sets of ifos, maybe want to be consistent here?

File: pycbc/events/coinc_rate.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    expected_coinc_rates[ifostring] = allowed_area * numpy.array(rateprod)
+    # if more than one possible coicidence type exists,
+    # calculate coincidences for subsets
```
Comment (tdent) [inline pycbc/events/coinc_rate.py]: subsets -> 'subsets via recursion' & remove later comment about calculate through iteration

File: pycbc/events/coinc_rate.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+            # add these sub-coincidences to the overall dictionary
+            for sub_coinc in sub_coinc_rates:
+                expected_coinc_rates[sub_coinc] = sub_coinc_rates[sub_coinc]
+
+    return expected_coinc_rates
+
+
+def multiifo_noise_coincident_area(ifos, slop):
+    """
+    calculate the multiplicative factor of the individual detector trigger
```
Comment (tdent) [inline pycbc/events/coinc_rate.py]: prefer to say what calculation the function is doing. eg 'total extent of time offset between 2 detectors, or area of the 2d space of time offsets for 3 detectors, for which a coinc can be generated'

File: pycbc/events/coinc_rate.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+            ifo2_num.append(int(numpy.mod(i+1, n_ifos)))
+            det0 = dets[ifo]
+            det1 = dets[ifos[ifo2_num[i]]]
+            tofs[i] = det0.light_travel_time_to_detector(det1) + slop
+
+        # combine these to calculate allowed area
+        allowed_area = 0
+        for i, _ in enumerate(ifos):
+            allowed_area += 2*tofs[i]*tofs[ifo2_num[i]] - tofs[i]**2
+
```
Comment (tdent) [inline pycbc/events/coinc_rate.py]: probably want to put something like `else raise NotImplementedError("Not able to deal with more than 3 ifos")` here (and remove the 'todo' about this)

File: pycbc/events/coinc_rate.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        tofs = numpy.zeros(n_ifos)
+        ifo2_num = []
+        # set up detector objects
+        for ifo in ifos:
+            dets[ifo] = pycbc.detector.Detector(ifo)
+
+        # calculate travel time between detectors (plus extra for timing error)
+        # TO DO: allow for different timing errors between different detectors
+        for i, ifo in enumerate(ifos):
+            ifo2_num.append(int(numpy.mod(i+1, n_ifos)))
```
Comment (tdent) [inline pycbc/events/coinc_rate.py]: PEP8, put spaces around operators. this also applies to line 110 (the actual area calculation)

File: pycbc/events/coinc_rate.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        allowed_area = 2*det0.light_travel_time_to_detector(det1)
+    elif n_ifos == 3:
+        dets = {}
+        tofs = numpy.zeros(n_ifos)
+        ifo2_num = []
+        # set up detector objects
+        for ifo in ifos:
+            dets[ifo] = pycbc.detector.Detector(ifo)
+
+        # calculate travel time between detectors (plus extra for timing error)
```
Comment (tdent) [inline pycbc/events/coinc_rate.py]: does this include timing error? doesn't seem like it does.

File: pycbc/events/coinc_rate.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        # TO DO: allow for different timing errors between different detectors
+        for i, ifo in enumerate(ifos):
+            ifo2_num.append(int(numpy.mod(i+1, n_ifos)))
+            det0 = dets[ifo]
+            det1 = dets[ifos[ifo2_num[i]]]
+            tofs[i] = det0.light_travel_time_to_detector(det1)
+
+        # combine these to calculate allowed area
+        phi_12 = numpy.arccos((tofs[0]**2+tofs[1]**2 -
```
Comment (tdent) [inline pycbc/events/coinc_rate.py]: spaces around operators please

File: pycbc/events/coinc_rate.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    for ifo in ifos:
+        dets[ifo] = pycbc.detector.Detector(ifo)
+    n_ifos = len(ifos)
+    if n_ifos == 2:
+        allowed_area = 2 * (dets[ifos[0]].light_travel_time_to_detector(
+                                                        dets[ifos[1]]) + slop)
+    elif n_ifos == 3:
+        tofs = numpy.zeros(n_ifos)
+        ifo2_num = []
+        # set up detector objects
```
Comment (tdent) [inline pycbc/events/coinc_rate.py]: did some code get eaten at this point?

Comment (GarethDaviesGW) [reply pycbc/events/coinc_rate.py]: I moved the creation of dets (and deprecated det1 and det0) to created outside of the if statement, which meant that the 2-ifo case could become a single line

Comment (tdent) [reply pycbc/events/coinc_rate.py]: ok, so fix the comment

File: pycbc/events/coinc_rate.py (hunk 5 above)
Comment (GarethDaviesGW) [reply pycbc/events/coinc_rate.py]: updated

**Comment by tdent** [comment]
@ahnitz I am OK with merging this - if you had any comments please speak soon.

Note that the formulae it implements will be (or are already - Gareth?) in the main text or appendix of the multi-ifo methods paper, comments pointing to them can be put in as and when published.

**Pull request merged**